### PR TITLE
[RFC] [AIR] Update `DLPredictor` interface

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

## Why are these changes needed?

**tl;dr** The `DLPredictor` interface is inflexible, and you have to override private methods to implement subclasses. This PR updates the interface to address the aforementioned problems.

### Change 1: Make `DLPredictor` and its required methods public

If you want to subclass `DLPredictor`, you need to override the private `_arrays_to_tensors`  and `_tensor_to_array` methods. You also need to import `DLPredictor` from the internal `ray.train._internal` subpackage.

**Before**
```python
from ray.train._internal.dl_predictor import DLPredictor

class MXNetPredictor(DLPredictor):
    def _arrays_to_tensors(...):
        ...
    def _tensor_to_array(...):
        ...
```

**After**
```python
from ray.train.dl_predictor import DLPredictor

class MXNetPredictor(DLPredictor):
    def array_to_tensor(...):  # This is now marked as a `DeveloperAPI`
        ...
    def tensor_to_array(...):  # This is now marked as a `DeveloperAPI`
        ...
```
   

### Change 2: Simplify `_arrays_to_tensors` interface 

`_arrays_to_tensors` requires boilerplate to handle `dict`s. We can simplify `DLPredictor` implementations by handling 
`dict`s elsewhere.

**Before**
```python
def _arrays_to_tensor(
    self,
    numpy_arrays: Union[np.ndarray, Dict[str, np.ndarray]],
    dtype: Union[np.dtype, Dict[str, np.dtype]],
) -> Union[mx.ndarray, Dict[str, mx.ndarray]]:
    if isinstance(numpy_arrays, dict):
        return {key, mx.ndarray(array) for key, array in numpy_arrays.items()}
    else:
        return mx.ndarray(numpy_arrays)
```

**After**
```python
def array_to_tensor(
    self,
    array: np.ndarray,
    dtype: np.dtype,
) -> mx.ndarray:
    return mx.ndarray(array)
```

### Change 3: Add `arrays_to_inputs` and `outputs_to_arrays`

The current `DLPredictor` interface assumes that models return tensors, but this isn't always the case. For example, Torch object detection models return lists of dicts. In this case, there's no easy way to override `DLPredictor`; you have to override `_predict_numpy`. 

`arrays_to_inputs` and `outputs_to_arrays` address this issue by giving you the flexibility to handle non-standard model inputs and outputs.  

**Before**
```python
class DLPredictor(Predictor):
    ...

    def _predict_numpy(
        self,
        data: Union[np.ndarray, Dict[str, np.ndarray]],
        dtype: Union[TensorDtype, Dict[str, TensorDtype]],
    ) -> Union[np.ndarray, Dict[str, np.ndarray]]:
        if isinstance(data, dict) and len(data) == 1:
            data = next(iter(data.values()))
        model_input = self._arrays_to_tensors(data, dtype)

        model_output = self.call_model(model_input)

        # NOTE: We're assuming that the model outputs a tensor or a dict that maps
        # strings to tensors.
        if isinstance(model_output, dict):
            return {k: self._tensor_to_array(v) for k, v in model_output.items()}
        else:
            return {"predictions": self._tensor_to_array(model_output)}
```

**After**
```python
class DLPredictor(Predictor):
    ...

    @DeveloperAPI
    def arrays_to_inputs(self, arrays: Union[np.ndarray, Dict[str, np.ndarray]]):
        ...  # Default implementation

    @DeveloperAPI
    def outputs_to_arrays(
        self, outputs: Any
    ) -> Union[np.ndarray, Dict[str, np.ndarray]]:
        ...  # Default implementation

    def _predict_numpy(
        self,
        data: Union[np.ndarray, Dict[str, np.ndarray]],
        dtype: Union[TensorDtype, Dict[str, TensorDtype]],
    ) -> Union[np.ndarray, Dict[str, np.ndarray]]:
        # NOTE: We make the same assumptions here, but now subclasses can 
        # override `arrays_to_inputs` and `outputs_to_arrays` if our assumptions
        # don't hold.
        model_inputs = self.arrays_to_inputs(data)
        model_outputs = self.call_model(model_inputs)
        return self.outputs_to_arrays(model_outputs)
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
Thanks for putting this together!

**Change 1**: The original intention for `DLPredictor` is for it not to be a user or developer facing API and to have it as as an internal abstraction to avoid duplicate code shared across `TorchPredictor` and `TensorflowPredictor`. To me, it would make sense to keep it this way to avoid introducing yet another developer API for implementing predictors. We already have two approaches: a lightweight approach via overriding `call_model` in `TorchPredictor` or `TensorflowPredictor` (see the `call_model` docstring for an example), and a more heavyweight one by overriding `_predict_numpy` or `_predict_pandas` in the base `Predictor` class which gives you full flexibility. Introducing a third approach in between these two doesn't seem to provide much value, and only adds to developer confusion IMO.

**Change 2**: This seems like a good change to make to simplify the API and avoid boilerplate code.

**Change 3**: Not sure I understand the value that this is providing- isn't overriding `arrays_to_inputs`, `call_model`, and `outputs_to_arrays` in your example exactly equivalent to just overriding `_predict_numpy`? Per discussion with @jiaodong previously, one potential API change needed here is to provide an escape hatch directly in `call_model`. Like this object detection predictor for example, which extends `call_model` to allow for pandas output directly: https://github.com/waleedkadous/movie-recs/blob/main/util.py#L61-L70. However, now with numpy narrow waist predictor, pandas dataframe output type may not be the best escape hatch.
